text/0000-computed-context.md

```diff
@@ -0,0 +1,329 @@
+
+**Specifying computed context**
+
+```ts
+const machine = createMachine({
+  context: {
+    firstName: '',
+    lastName: ''
+  },
+  computed: {
+    // Computed from non-computed context
+    fullName: ({ context }) => `${context.firstName} ${context.lastName}`,
+    // Computed from computed context
+    isLongName: ({ context }) => context.fullName.length > 20,
+    // Computed from state and context
+    errorMessage: (state) => {
+      if (state.hasTag('error')) {
+        return state.context.fullName.trim().length === 0
+          ? 'Invalid name'
+          : 'Unknown error'
+      }
+
+      return null;
+    },
+    // Returns a function, e.g.:
+    // `const greeting = state.context.greet('Hello')`
+    greet: ({ context }) => (greeting: string) => `${greeting}, ${context.fullName}`
+  }
+});
+```
+
+**Reading computed context**
+
+```js
+// ...
+
+const name = state.context.fullName; // computed from context.firstName and context.lastName
+
+const greeting = state.context.greet('Hello');
+```
+
+**Usage in assign**
+
+```ts
+const machine = createMachine({
+  context: {
+    firstName: '',
+    lastName: '',
+    age: 0
+  },
+  computed: {
+    fullName: ({ context }) => `${context.firstName} ${context.lastName}`,
+    // ...
+  },
+  // ...
+  entry: assign({
+    firstName: (_, event) => event.value,
+    age: (context, event) => {
+      // This will be the full name *prior* to the assignment above
+      // since it is within the same assign() call
+      console.log(context.fullName);
+
+      return event.value;
+    }
+  }),
+  // ...
+  exit: [
+    assign({ firstName: (_, event) => event.value }),
+    assign({
+      age: (context, event) => {
+        // This will be the full name *after* the previous assignment
+        // since it is a subsequent assign() call
+        console.log(context.fullName);
+
+        return event.value;
+      }
+    })
+  ],
+  // ...
+  on: {
+    someEvent: {
+      actions: assign({
+        // ❌ This will be a type error
+        fullName: 'something else'
+      })
+    }
+  }
+});
+```
+
+## How we teach this
+
+- We add a new "Computed Context" section to the existing guide on machine context
+- Computed context is introduced as "read-only context" where the computed values are determined from a function that takes in the `state` and returns the value.
+- Teach that computed context should not be assigned to
+- Teach that a computed context value can be a function, and show examples where this is useful
+- Demonstrate when computed context should and should not be used:
+  - Computed context should be used in place of normal context when a context value can be computed from other context values
+
+## Drawbacks
+
+- "I'd personally want a stronger reason than "convenience" for expanding the API (ie. adding to the amount of stuff I need to learn)" ([Discord](https://discord.com/channels/795785288994652170/977959576152449034/978117164697534485))
```

Comment:
This is both good and bad feedback. Good because it warns about expanding the API for a feature that might be used by every single user of the API.
Bad because it shows this feature can be implemented using an abstraction such `createMachineWithComputedValues` which is made on top of the built-in `createMachine`.